### PR TITLE
Add SocketConnectionFactory and http2cat

### PR DESCRIPTION
**Pull request by halter73**
- Add a Kestrel sample client for testing HTTP/2 servers
- Add SocketConnectionFactory

TODO: Add SslStream
TODO: Support half-closing the SocketConnection

### CODE REVIEW

File: src/Servers/Kestrel/samples/http2cat/Http2Utilities.cs [hunk 1]
```diff
@@ -0,0 +1,920 @@
...
+using Microsoft.Net.Http.Headers;
+
+namespace http2cat
+{
+    public class Http2Utilities : IHttpHeadersHandler
+    {
+        public static readonly int MaxRequestHeaderFieldSize = 16 * 1024;
+        public static readonly string _4kHeaderValue = new string('a', 4096);
+
+        public static readonly IEnumerable<KeyValuePair<string, string>> _browserRequestHeaders = new[]
```
Comment (jkotalik) [inline src/Servers/Kestrel/samples/http2cat/Http2Utilities.cs]: Anyway to share with this with our test infrastructure.

Comment (halter73) [reply src/Servers/Kestrel/samples/http2cat/Http2Utilities.cs]: Maybe. It's already diverged a fair bit from Http2TestBase, so I'm not sure it's worth the effort though.

File: src/Servers/Kestrel/samples/http2cat/http2cat.csproj [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp5.0</TargetFramework>
```
Comment (jkotalik) [inline src/Servers/Kestrel/samples/http2cat/http2cat.csproj]: This should be $(DefaultNetCoreTargetFramework)

**Review by jkotalik — APPROVED** [review]

**Comment by Tratcher** [comment]
This certainly meets its manual testing objectives. Here are some ideas for making it more broadly applicable in the future. E.g. I'd like to directly consume this as a library in Kestrel and HttpSys tests and remove the duplicate infrastructure.
- Move the bulk of the code from the sample to a class library so other components can easily depend on it. Only Program.cs would stay in the sample.
- Remove the internals-visible-to requirement, even if it requires copying more code out of Kestrel. Alternatively, those internals could be moved to a shared sources project that Kestrel directly consumes. Some of these internals overlap with the code we want to share with HttpClient anyways.

None of this is a near term priority.

**Comment by davidfowl** [comment]
@Tratcher can you capture that in an issue so we don’t lose it

**Pull request merged**